### PR TITLE
Fixes #1619 Too many log generation - LoadBalancer Heartbeat

### LINKED ISSUE #1619: Too many log generation - LoadBalancer Heartbeat
See [[vertx:36914] Too many log generation - LoadBalancer Heartbeat](https://groups.google.com/forum/#!topic/vertx/CF1qTX3Ak1E)

Connection is reset by client before we could determine if it's HTTP 1 or 2.

### PR DESCRIPTION
**Pull request by tsegismont**
Override exceptionCaught, log at trace level and close channel

### CODE REVIEW

File: src/test/java/io/vertx/test/core/HttpConnectionEarlyResetTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  private CountDownLatch resetLatch = new CountDownLatch(1);
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    CountDownLatch listenLatch = new CountDownLatch(1);
+    httpServer = new HttpServerImpl((VertxInternal) vertx, new HttpServerOptions()) {
+      @Override
+      protected void onChannelInitialized(Channel ch) {
+        ch.pipeline().addFirst(new ResetLatchCountDown()).addLast(new ThrowableRecorder());
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/HttpConnectionEarlyResetTest.java]: can we test without having to mess with Netty pipeline ?

Comment (vietj) [inline src/test/java/io/vertx/test/core/HttpConnectionEarlyResetTest.java]: I would rather be in favor of adding a connectionExceptionHandler in HttpServerImpl (not exposed) that would be called by the pipeline, the default implementation closes the channel.

The test can then change this handler to be aware of the failure and test properly.

Comment (tsegismont) [inline src/test/java/io/vertx/test/core/HttpConnectionEarlyResetTest.java]: Just pushed another version. Tell me if this is what you had in mind.

2016-09-13 14:03 GMT+02:00 Julien Viet <notifications@github.com>:

File: src/main/java/io/vertx/core/http/impl/HttpServerImpl.java [hunk 2]
```diff
@@ -304,6 +340,13 @@ protected void configurePipeline(ChannelHandlerContext ctx, String protocol) thr
     return this;
   }
 
+  // Visible for testing
+  protected void handleConnectionException(ChannelHandlerContext ctx, Throwable cause) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/HttpServerImpl.java]: ok so more like:

1. the close is done in Netty's handler
2. not a method to override but instead an `Handler<Throwable>` called `connectionExceptionHandler` that by default logs the cause and can be `set` in the test

File: src/main/java/io/vertx/core/http/impl/HttpServerImpl.java [hunk 3]
```diff
@@ -1019,5 +1063,11 @@ private void http1(ChannelHandlerContext ctx, ByteBuf buf) {
       ctx.fireChannelRead(buf);
       ctx.pipeline().remove(this);
     }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+      ctx.channel().close();
+      HttpServerImpl.this.connectionExceptionHandler.handle(cause);
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/HttpServerImpl.java]: I think it should use Context#executeFromIO and also test handler is not null.

Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/HttpServerImpl.java]: PTAL

2016-09-13 18:24 GMT+02:00 Julien Viet <notifications@github.com>:

File: src/main/java/io/vertx/core/http/impl/HttpServerImpl.java [hunk 4]
```diff
@@ -1019,5 +1065,13 @@ private void http1(ChannelHandlerContext ctx, ByteBuf buf) {
       ctx.fireChannelRead(buf);
       ctx.pipeline().remove(this);
     }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+      Channel channel = ctx.channel();
+      HandlerHolder<HttpHandler> reqHandler = reqHandlerManager.chooseHandler(channel.eventLoop());
+      reqHandler.context.executeFromIO(channel::close);
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/HttpServerImpl.java]: actually channel.close should be called normally but it's the callback to connectionExceptionHandler that should be executed with executeFromIO.

Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/HttpServerImpl.java]: Done

2016-09-14 18:25 GMT+02:00 Julien Viet <notifications@github.com>:

**Comment by vietj** [comment]
don't we have the same problem with the ApplicationProtocolNegotiationHandler ?

**Comment by tsegismont** [comment]
No, ApplicationProtocolNegotiationHandler has an exceptionCaught method
(which closes the channel).

2016-09-15 10:41 GMT+02:00 Julien Viet <notifications@github.com>:
